aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -2045,6 +2045,10 @@ Tensor all(const Tensor& self, Dimname dim, bool keepdim) {
 Tensor& all_out(const Tensor &self, Dimname dim, bool keepdim, Tensor& result) {
   reportNYIDimnameOverload("all");
 }
+Tensor _is_all_true(const Tensor& self) {
+  TORCH_INTERNAL_ASSERT(self.scalar_type() == at::kBool);
+  return self.all().unsqueeze(0);
```

Comment:
Evidently vmap only supports operations that return tensors, so I couldn't return a bool here.

Also, I couldn't figure out how to get the vmapped `_is_all_true` to return a scalar tensor. When `_is_all_true_batch_rule` returned a scalar, it would automatically get expanded into the batch size. So something like `vmap(torch._is_all_true)(torch.tensor([True, False]))`, for example, would return `tensor([False, False])`, even though the actual returned value of `_is_all_true_batch_rule` was just `tensor(False)`. But when I unsqueeze the result so that it's a 1-D tensor with one element, it doesn't get automatically resized. So there must be some weird special case for scalars, not sure if it's intentional or not

Reply:
There are two things going on here:
1. Ignoring the vmap problem, should `_is_all_true` return a bool or a Scalar Tensor? Tensor return seems better because it is more general, but curious to hear your thoughts.
2. The weird behavior you're seeing (vmap requiring Tensor returns, vmap expanding the result of things) is expected. Although, the thing we're trying to do is to change the return of `torch._is_all_true` inside of vmap, not necessarily return it as an output of vmap.

Let me expand a bit more on (2). The thing we want to work isn't exactly `vmap(torch._is_all_true)(torch.tensor([True, False]))`, it is the following:
```
def f(x):
  value = torch.is_all_true(x)
  assert value is False  # or True, or a Scalar Tensor
  return x
```
We're trying to let people perform error checks in an operation inside of vmap, not necessarily return the value of torch.is_all_true as an output from vmap.

Next, vmap has an out_dims argument that specifies where the vmapped dimension will be on the output Tensor. If the returned Tensor doesn't actually have the vmapped dimension, it gets auto-expanded. For example:
```py
def f(x):
  return torch.tensor(1.)  # scalar Tensor

x = torch.randn(3)  # vmap dimension of size 3
result = vmap(f)(x)
result  # torch.tensor(1., 1., 1.)
```

If we go the "`_is_all_true` returns Scalar Tensor" route, we want to check that *inside* vmap(f), the return of `_is_all_true` is a Scalar Tensor

Reply:
Ok those are good points. However, if I try to do a conditional using `_is_all_true()` in a vmapped function, I get this:

```python
RuntimeError: vmap: It looks like you're attempting to use a Tensor in some data-dependent control flow. We don't support that yet, please shout over at https://github.com/pytorch/functorch/issues/257 .
```

So evidently, a solution for that issue will need to be implemented before we can really do this.

At least this PR still accomplishes its main goal--to offer `TORCH_CHECK_TENSOR` that supports batched inputs. But I still need to write some kind of test for it